pkg/resource/v1/chartconfig/desired.go

```diff
@@ -2,8 +2,70 @@ package chartconfig
 
 import (
 	"context"
+
+	"github.com/giantswarm/apiextensions/pkg/apis/core/v1alpha1"
+	"github.com/giantswarm/microerror"
+	apimetav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/giantswarm/cluster-operator/pkg/cluster"
+	"github.com/giantswarm/cluster-operator/pkg/label"
+)
+
+const (
+	chartConfigAPIVersion           = "core.giantswarm.io"
+	chartConfigKind                 = "ChartConfig"
+	chartConfigVersionBundleVersion = "0.1.0"
+
+	chartOperatorChart           = "quay.io/giantswarm/chart-operator-chart"
+	chartOperatorChartConfigName = "chart-operator"
+	chartOperatorChannel         = "stable"
+	chartOperatorRelease         = "chart-operator"
 )
 
+// GetDesiredState returns all desired ChartConfigs for managed guest resources.
 func (r *Resource) GetDesiredState(ctx context.Context, obj interface{}) (interface{}, error) {
-	return nil, nil
+	clusterGuestConfig, err := r.toClusterGuestConfigFunc(obj)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	clusterConfig, err := prepareClusterConfig(*r.baseClusterConfig, *clusterGuestConfig)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	desiredChartConfigs := make([]*v1alpha1.ChartConfig, 0)
+	{
+		chartOperatorChartConfig := newChartOperatorChartConfig(clusterConfig, r.projectName)
+		desiredChartConfigs = append(desiredChartConfigs, chartOperatorChartConfig)
+	}
+
+	return desiredChartConfigs, nil
+}
+
+func newChartOperatorChartConfig(clusterConfig *cluster.Config, projectName string) *v1alpha1.ChartConfig {
+	return &v1alpha1.ChartConfig{
+		TypeMeta: apimetav1.TypeMeta{
+			Kind:       chartConfigKind,
+			APIVersion: chartConfigAPIVersion,
+		},
+		ObjectMeta: apimetav1.ObjectMeta{
+			Name: chartOperatorChartConfigName,
+			Labels: map[string]string{
+				label.Cluster:      clusterConfig.ClusterID,
+				label.ManagedBy:    projectName,
+				label.Organization: clusterConfig.Organization,
+			},
+		},
+		Spec: v1alpha1.ChartConfigSpec{
+			Chart: v1alpha1.ChartConfigSpecChart{
+				Name:    chartOperatorChart,
+				Channel: chartOperatorChannel,
+				Release: chartOperatorRelease,
+			},
+			VersionBundle: v1alpha1.ChartConfigSpecVersionBundle{
+				Version: chartConfigVersionBundleVersion,
+			},
+		},
+	}
```

Comment:
Thinking loud here if it wouldn't be better to create ChartConfigs by {aws,azure,kvm}-operators on their own as this would make them easier to test and more independent.

Reply:
I prefer doing this in cluster-operator because it needs to connect into guest clusters. Otherwise all 3 operators would need to do this.

I think this is still easy to test in cluster-operator. But it's a good point that there will be per platform differences in the desired state.

Reply:
They will need to connect anyways in scope of events story, also for filling status.

Reply:
That makes sense, but please discuss that. I don't think the downside is big here, especially other parts are managed by operators (e.g. OS, cloud config). Also you'll have to update the changelog. Also I like the fact operator decides on components it's installing. That's one of the problems with current k8scloudconfig we have. It's difficult to test. It's like moving part of logic from  {aws,azure,kvm}-operator to cluster-operator.

Reply:
I don't follow the argument here. We basically just make kubernetesd reconcile. Why should we move this centralized logic to multiple places? I think this here is right.

Reply:
I think it's because some of the guest components will differ by provider. For example on AWS we might create the Ingress Controller ELB as a K8s Load Balancer service rather than using aws-operator.

The downside for me is for shared components like node-exporter we need 3 PRs to update the version. Anyway the discussion ticket is here.

https://github.com/giantswarm/giantswarm/issues/2908

Reply:
> Why should we move this centralized logic to multiple places?

You can take this further and ask why we have separate operators for aws and azure. Of course benefit here is more obvious. I understand it as the same decision with more subtle reasons.

Reply:
The benefit for me is not obvious at all. And yes. We should actually have an API and cluster-service and everything per provider because of subtle differences everywhere but we have to cut it somewhere and the cut right now is just fine IMO.

Reply:
Other strange thing is so far we created CertConfigs and certs were used by {aws,azure,kvm}-operator. Then the operator creates cluster. Now creating cluster is distributed between cluster-operator and {aws,azure,kvm}-operator. For me the cut you describe isn't clear.

It's difficult for me to digest {aws,azure,kvm}-operator won't create clusters on their own. They'll need an operator above them. But maybe making cluster-operator our primary interface is the thing. Let's discuss.

Reply:
I am not sure why we even have this discussion. Nothing changes. Everything is designed as it was before. Yes, the design might lack and could be improved but what we are doing here is making the current design more reliable. We can make it nicer once we sorted real problems.

Reply:
chart-operator is new. If it makes sense we should make it nice from the beggining.